pkg/models/execution.go

```diff
@@ -0,0 +1,126 @@
+//go:generate stringer -type=ExecutionStateType --trimprefix=ExecutionState --output execution_state_string.go
+//go:generate stringer -type=ExecutionDesiredStateType --trimprefix=ExecutionDesiredState --output execution_desired_state_string.go
+package models
+
+// ExecutionStateType The state of an execution. An execution represents a single attempt to execute a job on a node.
+// A compute node can have multiple executions for the same job due to retries, but there can only be a single active execution
+// per node at any given time.
+type ExecutionStateType int
+
+// TODO: change states to reflect non-bidding scheduling
```

Comment:
When this happens will we replace the bid based states with New/Pending/Running? 
I'm keen for us to try and differentiate between "Preparing the environment for the run" and "The docker image is actually starting".

Reply:
I am still planning to keep the two phase of job scheduling where the scheduler can ask multiple nodes to run a job, but only select few at the end. I am planning to change the language and order of events so that we end up with a [Sparrow](https://cs.stanford.edu/~matei/papers/2013/sosp_sparrow.pdf) implementation for job scheduling. It will look as follows:
1. The scheduler will submit a job to multiple nodes. If they can run the job, they will enqueue it, otherwise they will fail the request. This is similar to `AskForBid`, but I am hoping to make the response synchronous to simplify things. Will need to sync up with @simonwo about this. The other difference is the compute nodes will enqueue the job right away and won't wait for approval
2. When the compute node is ready to execute the job, it will first ask the scheduler for approval in case another compute node already started the execution. If this is the first node that is ready to execute, then the scheduler will approve the request, send rejection to all other compute nodes, and will reject any future request to approve the execution in case the compute node didn't hear about the rejection yet
3.  This two phase job scheduling will only be used for `Batch` and `Service` jobs, but not for `Daemon` or `Ops` jobs

Reply:
The reason we have async bid approval is so that the external approval hooks (HTTP and exec) can take a long time to approve things (e.g. many seconds to do a network request) if they want to. I think that is generally a positive quality of life improvement that we should not remove.